runtime/runtime/src/ethereum.rs

```diff
@@ -0,0 +1,162 @@
+#![allow(unused)]
+use std::mem;
+
+use bigint::{H256, H64, U256};
+use std::fs::{create_dir_all, File};
+use std::io::{self, Read, Write};
+use std::path::{Path, PathBuf};
+
+const ETHASH_EPOCH_LENGTH: u64 = 30000;
+const NODE_BYTES: usize = 64;
+
+pub struct LightCache {
+    epoch: u64,
+    cache: Vec<u8>,
+}
+
+impl LightCache {
+    fn storage_file(cache_dir: &PathBuf, epoch: u64) -> PathBuf {
+        let mut path = cache_dir.clone();
+        path.push(format!("{}", epoch));
+        path
+    }
+
+    pub fn new(cache_dir: &PathBuf, epoch: u64) -> Self {
+        match Self::from_file(cache_dir, epoch) {
+            Ok(light_cache) => light_cache,
+            Err(_) => {
+                let cache_size = ethash::get_cache_size(epoch as usize);
+                let mut cache = Vec::with_capacity(cache_size);
+                cache.resize(cache_size, 0);
+                let seed = ethash::get_seedhash(epoch as usize);
+                ethash::make_cache(&mut cache, seed);
+                let _ = Self::to_file(cache_dir, epoch, &cache);
+                LightCache { epoch, cache }
+            }
+        }
+    }
+
+    fn to_file(cache_dir: &PathBuf, epoch: u64, cache: &[u8]) -> io::Result<usize> {
+        create_dir_all(cache_dir)?;
+        let mut file = File::create(Self::storage_file(cache_dir, epoch))?;
+        file.write(cache)
+    }
+
+    pub fn from_file(cache_dir: &PathBuf, epoch: u64) -> io::Result<Self> {
+        let mut file = File::open(Self::storage_file(cache_dir, epoch))?;
```

Comment:
It doesn't provide much protection on same file reading/writing

Reply:
What's the better option?